gcp-deployer/deploy/ssh.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deploy
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/golang/glog"
+)
+
+const (
+	MachineControllerSshKeySecret = "machine-controller-sshkeys"
+
+	// Arbitrary name used for SSH.
+	SshUser                = "clusterapi"
+	SshUserFile            = "clusterapi-user"
+	SshKeyFilePrivate      = "clusterapi-key"
+	SshKeyFilePublic       = SshKeyFilePrivate + ".pub"
+)
+
+func createSshKeyPairs() error {
+	if fileExists(SshUserFile) || fileExists(SshKeyFilePrivate) || fileExists (SshKeyFilePublic) {
+		if !fileExists(SshUserFile) || !fileExists(SshKeyFilePrivate) || !fileExists (SshKeyFilePublic) {
+			return fmt.Errorf(
+				"A incomplete set of ssh files exist and cannot be reused. Please remove all files to regenerate ssh keys: %v, %v, %v",
+				SshUserFile,
+				SshKeyFilePrivate,
+				SshKeyFilePublic)
+		} else {
+			glog.Info("Re-using existing ssh files.\n")
+		}
+	}
+
+	_, err := exec.Command("ssh-keygen", "-t", "rsa", "-f", SshKeyFilePrivate, "-C", SshUser, "-N", "").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("couldn't generate RSA keys: %v", err)
+	}
+
+	f, err := os.Create(SshUserFile)
+	if err != nil {
+		return fmt.Errorf("couldn't create ssh user file: %v", err)
+	}
+
+	_, err = f.WriteString(SshUser)
+	if err != nil {
+		return fmt.Errorf("couldn't write ssh user file: %v", err)
+	}
+
+	glog.Info("Created ssh files.\n")
+
+	return nil
+}
+
+func cleanupSshKeyPairs() {
+	os.Remove(SshKeyFilePrivate)
+	os.Remove(SshKeyFilePublic)
+	os.Remove(SshUserFile)
+}
+
+// It creates secret to store private key.
+func setupSSHSecret() error {
+	// Create secrets so that machine controller container can load them
+	c := exec.Command("kubectl",
+		"create",
+			"secret",
+			"generic",
+			MachineControllerSshKeySecret,
+			"--from-file=private="+SshKeyFilePrivate,
+		  "--from-file=public="+SshKeyFilePublic,
+			"--from-literal=user="+SshUser)
+	out, err := c.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("couldn't create service account key as credential. error: %v output: %v", err, string(out))
+	}
+
+	return err
+}
+
+func remoteSshCommand(ip, cmd, sshKeyPath, sshUser string) (string, error) {
+	glog.Infof("Remote SSH execution '%s' on %s", cmd, ip)
+
+	c := exec.Command("ssh", "-i", sshKeyPath, "-q", sshUser+"@"+ip, "-o", "StrictHostKeyChecking=no", cmd)
```

Comment:
Since we create the master, it will be an unknown host to the machine running ssh. This triggers an interactive prompt and that does not work for automation. I am under the impression I am not going to be able to avoid man in the middle attacks. However, I can reduce the surface of them by only having it on the first connection. 
Issue to track improvement: https://github.com/kubernetes-sigs/cluster-api/issues/187

Reply:
The "-q" flag prevents that warning (have recently tested and submitted a PR to that end). I am surprised that you are running into the interactive prompt. Have you tried it recently?